src/common/qat.c

```diff
@@ -0,0 +1,537 @@
+/**
+ * (C) Copyright 2019-2020 Intel Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. B609815.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+#define D_LOGFAC	DD_FAC(csum)
+
+#ifdef HAVE_QAT
+#include <gurt/types.h>
+#include <semaphore.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "cpa.h"
+#include "cpa_cy_im.h"
+#include "cpa_cy_sym.h"
+#include "icp_sal_user.h"
+#include "icp_sal_poll.h"
+#include "qae_mem.h"
+
+#define MAX_INSTANCES 32
+
+/* Use semaphores to signal completion of events */
+struct completion_struct
+{
+    sem_t semaphore;
+};
+
+#define COMPLETION_STRUCT completion_struct
+
+#define COMPLETION_INIT(s) sem_init(&((s)->semaphore), 0, 0);
+
+#define COMPLETION_WAIT(s, timeout) (sem_wait(&((s)->semaphore)) == 0)
+
+#define COMPLETE(s) sem_post(&((s)->semaphore))
+
+#define COMPLETION_DESTROY(s) sem_destroy(&((s)->semaphore))
+
+/* Memory */
+#define OS_MALLOC(ppMemAddr, sizeBytes)                                        \
+    Mem_OsMemAlloc((void *)(ppMemAddr), (sizeBytes))
+
+#define PHYS_CONTIG_ALLOC(ppMemAddr, sizeBytes)                                \
+    Mem_Alloc_Contig((void *)(ppMemAddr), (sizeBytes), 1)
+
+#define PHYS_CONTIG_ALLOC_ALIGNED(ppMemAddr, sizeBytes, alignment)             \
+    Mem_Alloc_Contig((void *)(ppMemAddr), (sizeBytes), (alignment))
+
+#define OS_FREE(pMemAddr) Mem_OsMemFree((void *)&pMemAddr)
+
+#define PHYS_CONTIG_FREE(pMemAddr) Mem_Free_Contig((void *)&pMemAddr)
+
+extern CpaStatus qaeMemInit(void);
+extern void qaeMemDestroy(void);
+
+/* Threads */
+typedef pthread_t sampleThread;
+static sampleThread gPollingThread;
+static volatile int gPollingCy = 0;
+
+static __inline CpaStatus Mem_OsMemAlloc(void **ppMemAddr, Cpa32U sizeBytes)
+{
+    *ppMemAddr = malloc(sizeBytes);
+    if (NULL == *ppMemAddr)
+    {
+        return CPA_STATUS_RESOURCE;
+    }
+    return CPA_STATUS_SUCCESS;
+}

```

Comment:
Could just use our D_ALLOC/FREE?

Reply:
Not sure. Maybe we can evaluate and do it in separate PR?

Reply:
Will D_ALLOC allocate physical continuous memory?

Reply:
No, but you are calling malloc above so it won't be physically contiguous anyways, no?

Reply:
OK, I thought it was Mem_Alloc_Contig, I've replaced the OS_MALLOC/FREE with D_ALLOC/FREE, will update in the new PR.